subsys/audio_module/audio_module.h

```diff
@@ -0,0 +1,463 @@
+
+/**
+ * @brief Number of valid least significant bits in the channel map
+ *
+ */
+#define AMOD_BITS_IN_CHANNEL_MAP (32)
+
+/**
+ * @brief Channel positions within the channel map
+ *
+ */
+#define AMOD_CHANNEL_LEFT_FRONT	 1
+#define AMOD_CHANNEL_RIGHT_FRONT 2
+#define AMOD_CHANNEL_CENTRE	 4
+#define AMOD_CHANNEL_LFE	 8
+#define AMOD_CHANNEL_LEFT_BACK	 16
+#define AMOD_CHANNEL_RIGHT_BACK	 32
+
+/**
+ * @brief Module type.
+ *
+ */
+enum audio_module_type {
+	/* The module type is undefined */
+	AMOD_TYPE_UNDEFINED = 0,
+
+	/* This is an input processing module.
+	 * Note: An input module obtains data internally within the
+	 *       module (e.g. I2S) and hence has no RX FIFO.
+	 */
+	AMOD_TYPE_INPUT,
+
+	/* This is an output processing module.
+	 * Note: An output module outputs data internally eithin the
+	 *       module (e.g. I2S) and hence has no TX FIFO.
+	 */
+	AMOD_TYPE_OUTPUT,
+
+	/* This is a processing module.
+	 * Note: An processing module takes input and outputs from/to another
+	 *       module, thus having RX and TX FIFOs.
+	 */
+	AMOD_TYPE_PROCESS
+};
+
+/**
+ * @brief Module state.
+ */
+enum audio_module_state {
+	/* Module state undefined */
+	AMOD_STATE_UNDEFINED = 0,
+
+	/* Module is in the configured state */
+	AMOD_STATE_CONFIGURED,
+
+	/* Module is in the running state */
+	AMOD_STATE_RUNNING,
+
+	/* Module is in the stopped state */
+	AMOD_STATE_STOPPED
+};
+
+/**
+ * @brief Module's private handle.
+ */
+struct audio_module_handle_private;
+
+/**
+ * @brief Private context for the module.
+ */
+struct audio_module_context;
+
+/**
+ * @brief Module's private configuration.
+ */
+struct audio_module_configuration;
+
+/**
+ * @brief Callback function for a response to a data_send as
+ *        supplied by the module user.
+ *
+ * @param handle[in/out]  The handle of the module that sent the data message
+ * @param block[in]       The audio block to operate on
+ */
+typedef void (*audio_module_response_cb)(struct audio_module_handle_private *handle,
+					 struct audio_data *block);
+
+/**
+ * @brief Private pointer to a module's functions.
+ */
+struct audio_module_functions {
+	/**
+	 * @brief Function for opening a module.
+	 *
+	 * @param handle[in/out]     The handle to the module instance
+	 * @param configuration[in]  Pointer to the module's current configuration
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*open)(struct audio_module_handle_private *handle,
+		    struct audio_module_configuration *configuration);
+
+	/**
+	 * @brief Function to close an open module.
+	 *
+	 * @param handle[in/out]  The handle to the module instance
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*close)(struct audio_module_handle_private *handle);
+
+	/**
+	 * @brief Function to configure a module.
+	 *
+	 * @param handle[in/out]     The handle to the module instance
+	 * @param configuration[in]  Pointer to the module's configuration to set
+	 *
+	 * @return 0 if successful, error otherwise
```

Comment:
Can we mention something about why we can/should configure here and also in the `open` function?